src/frontend/Typechecker.ml

```diff
@@ -1001,40 +1002,178 @@ let warn_self_assignment loc lhs rhs =
+    match (lv.lval, type_) with
+    | _, UTuple ts ->
+        List.concat_mapi ts ~f:(fun i ty ->
+            add_tuple_idxs
+              { lval= LTupleProjection (lv, i + 1)
+              ; lmeta= {lv.lmeta with type_= ty} } )
+    | _ -> [lv] in
+  let rec flatten_lvalue_pack lv =
+    match lv with
+    | LTuplePack (lvs, _) -> List.concat_map ~f:flatten_lvalue_pack lvs
+    | LValue lv -> [lv] in
+  flatten_lvalue_pack lv
+  |> List.concat_map ~f:add_tuple_idxs
+  |> List.map ~f:Ast.untyped_lvalue_of_typed_lvalue
+
+let variables_accessed_in lv =
+  (* We only care about values being read inside the lvalue here,
+     which means only the expressions inside of LIndexed *)
+  let rec extract_indices lv =
+    match lv.lval with
+    | LVariable _ -> []
+    | LTupleProjection (lv, _) -> extract_indices lv
+    | LIndexed (lv, es) -> extract_indices lv @ es in
+  let rec extract_indices_pack lv =
+    match lv with
+    | LTuplePack (lvs, _) -> List.concat_map ~f:extract_indices_pack lvs
+    | LValue lv -> extract_indices lv in
+  let exprs_in_index = function
+    | All -> []
+    | Single e -> [e]
+    | Upfrom e -> [e]
+    | Downfrom e -> [e]
+    | Between (e1, e2) -> [e1; e2] in
+  let all_variables =
+    lv |> extract_indices_pack
+    |> List.concat_map ~f:exprs_in_index
+    |> List.concat_map ~f:extract_ids
+    |> List.dedup_and_sort ~compare:Ast.compare_identifier in
+  all_variables
+
+(** When an lvalue is assigning to multiple places at once (e.g., a tuple
+    is being unpacked), we want to ensure that the same memory is not being
+    written to multiple times. This lets us preserve the illusion of
+    simultaneous assignment and generally avoid confusion.
+    For the same reasons, we also want to avoid reading the value of
+    a variable which is being updated in this same lvalue. *)
+let verify_lvalue_unique (lv : Ast.typed_lval_pack) =
+  let loc =
+    match lv with LValue {lmeta= {loc; _}; _} | LTuplePack (_, loc) -> loc in
+  let all_lvals = lvalues_written_to lv in
+  let () =
+    (* check that things being assigned to are all unique *)
+    match overlapping_lvalues all_lvals with
+    | [] -> ()
+    | dupes ->
+        Semantic_error.cannot_assign_duplicate_unpacking loc dupes |> error
+  in
+  (* check that things being assigned to are not also being read
+     Note: this is much less refined than the above and forbids some
+     cases that would be harmless, but this is also in general a very
+     weird thing to try to do, so I think that is acceptable
+  *)
+  let rec extract_id lv =
+    match lv.lval with
+    | LVariable id -> id
+    | LTupleProjection (lv, _) -> extract_id lv
+    | LIndexed (lv, _) -> extract_id lv in
+  let all_variables =
+    List.map ~f:extract_id all_lvals
+    |> List.dedup_and_sort ~compare:Ast.compare_identifier in
+  let accessed_lvals = variables_accessed_in lv in
+  match
+    List.find_all_dups ~compare:Ast.compare_identifier
+      (all_variables @ accessed_lvals)
```

Comment:
If `variables_accessed_in` produces a set like I suggested in the above comment, this simplifies
```suggestion
    List.filter ~f:(Set.mem accessed_lvals) all_variables
```